### PR TITLE
Eth ABI Decode support and some fixes

### PR DESCRIPTION
**Pull request by catenocrypt**
In preparation for Issue #573 :
There was Eth ABI encoding support, but not decoding.  Decoding added, along with tests.  For encoding coverage increased and few fixes added.

### CODE REVIEW

File: src/Ethereum/ABI/Bytes.h [hunk 1]
```diff
@@ -12,70 +12,87 @@
 
 namespace TW::Ethereum {
 
-// dynamic bytes
+/// Compute size when padded to 32-byte boundary; round up to multiple of 32.
+inline size_t paddedTo32(size_t len) { return ((len + 31) / 32) * 32; }
 
-bool is_dynamic(Data) {
+/// Compute padding needed to next 32-byte boundary.
+inline size_t padNeeded32(size_t len) { return ((len + 31) / 32) * 32 - len; }
```
Comment (hewigovens) [inline src/Ethereum/ABI/Bytes.h]: looks like we can add a `bool` parameter to  `paddedTo32` and merge these two functions?

File: src/Ethereum/ABI/Numbers.h [hunk 2]
```diff
@@ -15,132 +15,160 @@ static constexpr std::size_t encodedIntSize = 32;
...
 }
 
-void encode(int256_t value, Data& data) {
+inline void encode(int256_t value, Data& data) {
     encode(static_cast<uint256_t>(value), data);
 }
 
-std::string type_string(int256_t value) {
+template<typename T>
+inline bool decodeGenericNumber(const Data& encoded, T& decoded, size_t& offset_inout) {
```
Comment (hewigovens) [inline src/Ethereum/ABI/Numbers.h]: maybe just `decodeNumber`? T is already a generic type

File: src/Ethereum/ABI/Tuple.h [hunk 3]
```diff
@@ -89,25 +89,70 @@ void encode_dynamic(const std::tuple<First, T...>& tuple, Data& data) {
...
         string += "," + type_string(tail(tuple));
     }
     return string;
 }
 
+inline bool decodeTuple_small(const Data& encoded, std::tuple<>& tuple_inout, size_t& offset_inout) { return true; }
+
+/// Process small values and extract indices for larger values
+template <typename First, typename... T>
+inline bool decodeTuple_small(const Data& encoded, std::tuple<First, T...>& tuple_inout, size_t& offset_inout) {
```
Comment (hewigovens) [inline src/Ethereum/ABI/Tuple.h]: just naming, `decodeTupleSmall` and `decodeTupleLarge`

File: tests/Ethereum/ABITests.cpp [hunk 4]
```diff
@@ -53,7 +67,122 @@ TEST(ABI, EncodeNegativeBigInt) {
...
+}
+
+TEST(ABI, EncodeTupleMixed) {
+    Data encoded;
+    auto tuple = std::make_tuple(uint256_t(69u), std::vector<uint256_t>{1, 2, 3}, true, std::string("Hello"), Data{0x64, 0x61, 0x76, 0x65});
+    encode(tuple, encoded);
+
+    EXPECT_EQ(13 * 32, encoded.size());
+    EXPECT_EQ(
+        "0000000000000000000000000000000000000000000000000000000000000045"\
```
Comment (hewigovens) [inline tests/Ethereum/ABITests.cpp]: we don't need to add `\`

File: src/Ethereum/ABI/Numbers.h (hunk 2 above)
Comment (catenocrypt) [reply src/Ethereum/ABI/Numbers.h]: OK, accepted, thx

File: src/Ethereum/ABI/Tuple.h (hunk 3 above)
Comment (catenocrypt) [reply src/Ethereum/ABI/Tuple.h]: ok,accepted, thx

File: tests/Ethereum/ABITests.cpp (hunk 4 above)
Comment (catenocrypt) [reply tests/Ethereum/ABITests.cpp]: indeed...!

File: src/Ethereum/ABI/Bytes.h (hunk 1 above)
Comment (catenocrypt) [reply src/Ethereum/ABI/Bytes.h]: I think that would not make it more readable, I would not do.

**Pull request merged**